### ISSUE TITLE
Spread not highlighted correctly

### ISSUE THREAD

**Issue by texastoland** [nushell/vscode-nushell-lang#173]
Looking at the grammar file I assume it just hasn't been added yet. Minimal repro:

```
[...$stuff]
{ ...$stuff }
```

**Comment by fdncred** [comment]
Are these the only two variations? It would be good to have an example of each variation so they can be validated.

I can see that `{ ...$rest }` seems correct and is called `variable.other.nushell` but `[...$rest]` isn't recognized properly and it's either `string.bare.nushell` or `keyword.control.nushell` depending on where you click on the text.

@glcraft Thoughts on this?

**Comment by texastoland** [comment]
> I can see that `{ ...$rest }` seems correct

They highlight identically for me. The first `..` is highlighted like `match` then the rest like a bare identifier.

**Comment by fdncred** [comment]
Not identical for me.
![image](https://github.com/nushell/vscode-nushell-lang/assets/343840/f8d9bce9-31b1-48f2-957c-386dccc991f9)

**Comment by fdncred** [comment]
Now that I look closer, neither of them are recognized correctly but one is closer than the other.
![image](https://github.com/nushell/vscode-nushell-lang/assets/343840/3851c0f0-a9f7-4adc-8459-89a8a677f520)

You can see this menu with ctrl+shift+p `Developer: Inspect Editor Tokens and Scope`

**Comment by texastoland** [comment]
It looks like our themes target different scopes.
![Code 2024-02-15 at 14 59 47@2x](https://github.com/nushell/vscode-nushell-lang/assets/48431/4c7c3716-b1da-48e8-bb0d-79ec32f67f5c)

**Comment by texastoland** [comment]
Found it. `|\\.\\.\\.` just needs added to this line:

https://github.com/nushell/vscode-nushell-lang/blob/bd9d19fd83a010af6b007def456d59138d44a59f/syntaxes/nushell.tmLanguage.json#L195

PR I guess?

**Comment by fdncred** [comment]
@texastoland Are you on the latest version of the extension v1.8.0?

**Comment by texastoland** [comment]
I am. You'll see the same behavior if you inspect inside the operator. Your theme just just highlights fewer scopes.

The right fix looks adapting (and perhaps renaming) the `ranges` regex to permit a third dot:

https://github.com/nushell/vscode-nushell-lang/blob/bd9d19fd83a010af6b007def456d59138d44a59f/syntaxes/nushell.tmLanguage.json#L186-L189

To: `"\\.\\.[<.]?|:"` PR??

**Comment by fdncred** [comment]
What theme are you using? I'd like to be able to see the difference.

**Comment by texastoland** [comment]
Oh just checked my other themes but they're like yours. Screenshot used [Tinacious Design](https://marketplace.visualstudio.com/items?itemName=tinaciousdesign.theme-tinaciousdesign).

**Comment by fdncred** [comment]
That range fix may be helpful but I don't think it's the entire solution. I don't think we should be getting multiple primary textmate scopes as I move my cursor from left to right. I get these primary scopes.

- [ ...$rest ]
  - keyword.control.nushell
  - string.bare.nushell
- { ...$rest }
  - keyword.control.nushell
  - entity.name.type.external.nushell
  - variable.other.nushell

**Comment by texastoland** [comment]
> ```
> [...$stuff]
> { ...$stuff }
> ```

From my minimal example here's the (unfixed) scopes I see:

| | `[]` | `{}` |
| - | - | - |
| **`..`** | `keyword.control.nushell` | `keyword.control.nushell` |
| | `meta.table.nushell` | `meta.expression.braced.nushell` |
| | `source.nushell` | `source.nushell` |
| **`.$rest`** | `string.bare.nushell` | `variable.other.nushell` `meta.command.nushell` |
| | `meta.table.nushell` | `meta.expression.braced.nushell` |
| | `source.nushell` | `source.nushell` |

It looks my patch would end up interpret `...` as an operator inside a closure instead of spreading a record. The alternative would be to add a new `spread` expression to each of the `table` and `braced` groups:

https://github.com/nushell/vscode-nushell-lang/blob/bd9d19fd83a010af6b007def456d59138d44a59f/syntaxes/nushell.tmLanguage.json#L215-L221

https://github.com/nushell/vscode-nushell-lang/blob/bd9d19fd83a010af6b007def456d59138d44a59f/syntaxes/nushell.tmLanguage.json#L516-L565

**Comment by fdncred** [comment]
ya, I think we should have a new spread_expression to accomodate this new syntax.

**Comment by texastoland** [comment]
I'll PR tomorrow then and include a table with the new scopes 👌🏼

**Comment by texastoland** [comment]
I started implementing this and ran into a problem. There are a few more cases to support:

```nushell
def cmd [ ...$rest ] {
  [ ...$rest ] # list
  { ...$rest } # record
  all ...$rest # builtin command
}
cmd ...$rest # custom command
^cmd ...$rest # external command
```

The problem arises in the way external commands are parsed:

https://github.com/nushell/vscode-nushell-lang/blob/bd9d19fd83a010af6b007def456d59138d44a59f/syntaxes/nushell.tmLanguage.json#L648-L662

TLDR `command` is only the name then it goes on to parse any arbitrary expression. That means breaking out a `spread` subexpression will ultimately parse identically to just making `ranges` more general.

I see 3 solutions:

1. Reuse `ranges` (my preference)
2. Add a `spread` expression to the linked `patterns` above (also easy)
3. Significantly refactor `commands` to encompass the entire call (flags etc.)

**Comment by texastoland** [comment]
Deleted a previous comment because I was mistaken and amended my PR 🪚
